core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -145,15 +146,21 @@ class LogSegment(val log: FileRecords,
   private def updateProducerState(producerStateManager: ProducerStateManager, batch: RecordBatch): Unit = {
     if (batch.hasProducerId) {
       val producerId = batch.producerId
-      val appendInfo = producerStateManager.prepareUpdate(producerId, loadingFromLog = true)
-      val maybeCompletedTxn = appendInfo.append(batch)
-      producerStateManager.update(appendInfo)
-      maybeCompletedTxn.foreach { completedTxn =>
-        val lastStableOffset = producerStateManager.completeTxn(completedTxn)
-        updateTxnIndex(completedTxn, lastStableOffset)
+      val appendInfo = producerStateManager.prepareUpdate(producerId, acceptFirstAppendUnconditionally = true)
+      try {
+        val maybeCompletedTxn = appendInfo.append(batch)
+        producerStateManager.update(appendInfo)
+        maybeCompletedTxn.foreach { completedTxn =>
+          val lastStableOffset = producerStateManager.completeTxn(completedTxn)
+          updateTxnIndex(completedTxn, lastStableOffset)
+        }
+      } catch {
+        case _ : DuplicateSequenceException =>
+          info(s"Not updating producer state with producerID: ${batch.producerId()}, and base sequence ${batch.baseSequence()}, " +
```

Comment:
Nit: we typically write "producer id" in log statements. Also, `()` is not needed in either of the two variables.